ddraw/IDirect3DDeviceX.cpp

```diff
@@ -2164,6 +2208,26 @@ HRESULT m_IDirect3DDeviceX::DrawIndexedPrimitive(D3DPRIMITIVETYPE dptPrimitiveTy
 		}
 		else
 		{
+			const UINT stride = GetVertexStride(dwVertexTypeDesc);
+
+			// Handle PositionT
+			if((dwVertexTypeDesc & D3DFVF_XYZRHW) != 0 && Config.DdrawConvertHomogeneousW)
+			{
+				UINT8 *vertex = (UINT8*)lpVertices;
+
+				for (UINT x = 0; x < dwVertexCount; x++)
+				{
+					float *pos = (float*) vertex;
+
+					pos[3] = 1.0f;
```

Comment:
Don't you lose information here if rhw is != 1?
It's not used for many things, but at least for perspective-correct texture mapping and some exotic stuff like fog.

Reply:
The documentation read that the use of W is undefined here, and not seen as part of the position vector. For me it made no difference if I had it or not. So it is probably better to just remove it then.

Reply:
Yeah not sure if they use the W at all in this mode, also they also say it's "designed for, and can only be used with, the programmable vertex pipeline."
https://learn.microsoft.com/en-us/windows/win32/direct3d9/d3dfvf

But I still don't understand how this could possibly work, the input coordinates are in screen space and this mode expects clip space.

https://gamedev.net/forums/topic/582097-difference-between-d3dfvf_xyzw-and-d3dfvf_xyzrhw/4705397/

Reply:
The SetTransform function overrides the game matrices with its own, when the mode is active.

```
// Replace the matrix with one that handles D3DFVF_XYZRHW geometry
D3DVIEWPORT9 Viewport9;
if(SUCCEEDED((*d3d9Device)->GetViewport(&Viewport9)))
{
	ZeroMemory(lpD3DMatrix, sizeof(_D3DMATRIX));
	lpD3DMatrix->_11 = 2.0f / (float)Viewport9.Width;
	lpD3DMatrix->_22 = -2.0f / (float)Viewport9.Height;
	lpD3DMatrix->_33 = 1.0f;
	lpD3DMatrix->_41 = -1.0f;  // translate X
	lpD3DMatrix->_42 = 1.0f;   // translate Y
	lpD3DMatrix->_44 = 1.0f;
}
```

Reply:
I know.
I guess an identity viewport matrix would also suffice but not sure about this mode.
https://learn.microsoft.com/en-us/windows/win32/dxtecharts/the-direct3d-transformation-pipeline